File: android_native/app/src/main/java/com/familymotion/app/ui/viewmodel/ExerciseViewModel.kt
package com.familymotion.app.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.familymotion.app.data.model.*
import com.familymotion.app.data.repository.StandardActionRepository
import com.familymotion.app.domain.service.ActionScoringEngine
import com.familymotion.app.domain.service.PoseDetectionService
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * 运动状态
 */
data class ExerciseState(
    val plan: ExercisePlan? = null,
    val currentExerciseIndex: Int = 0,
    val isCountingDown: Boolean = false,
    val countdown: Int = 3,
    val isPlaying: Boolean = false,
    val currentScore: ScoreResult? = null,
    val repCount: Int = 0,
    val maxReps: Int = 0,
    val totalScore: Int = 0,
    val exerciseScores: List<Int> = emptyList(),
    val isCompleted: Boolean = false,
    val poseData: PoseData? = null
) {
    val currentExercise: Exercise?
        get() = plan?.exercises?.getOrNull(currentExerciseIndex)
}

/**
 * 运动 ViewModel
 */
class ExerciseViewModel(
    private val poseDetectionService: PoseDetectionService? = null
) : ViewModel() {

    private val _state = MutableStateFlow(ExerciseState())
    val state: StateFlow<ExerciseState> = _state.asStateFlow()

    private val scoringEngine = ActionScoringEngine()
    private var previousPoseData: PoseData? = null
    private var scoringJob: Job? = null

    /**
     * 初始化运动
     */
    fun initExercise(plan: ExercisePlan) {
        scoringEngine.reset()
        _state.value = ExerciseState(
            plan = plan,
            currentExerciseIndex = 0,
            maxReps = plan.exercises.firstOrNull()?.reps ?: 0
        )
    }

    /**
     * 开始倒计时
     */
    fun startCountdown() {
        _state.value = _state.value.copy(isCountingDown = true, countdown = 3)

        viewModelScope.launch {
            for (i in 3 downTo 1) {
                _state.value = _state.value.copy(countdown = i)
                delay(1000)
            }
            _state.value = _state.value.copy(isCountingDown = false)
            startExercise()
        }
    }

    /**
     * 开始运动
     */
    private fun startExercise() {
        _state.value = _state.value.copy(isPlaying = true)
        scoringJob = viewModelScope.launch {
            while (_state.value.isPlaying) {
                delay(500) // 每500ms评分一次

                val poseData = _state.value.poseData
                val exercise = _state.value.currentExercise

                if (poseData != null && exercise != null) {
                    // 评分
                    val scoreResult = scoringEngine.scoreAction(poseData, exercise)
                    _state.value = _state.value.copy(
                        currentScore = scoreResult,
                        repCount = scoreResult.repCount
                    )

                    // 检测动作完成
                    if (scoringEngine.detectRepCompletion(poseData, previousPoseData, exercise)) {
                        handleRepCompleted()
                    }

                    previousPoseData = poseData
                }
            }
        }
    }

    /**
     * 处理动作完成
     */
    private fun handleRepCompleted() {
        val currentState = _state.value
        val newRepCount = currentState.repCount + 1

        _state.value = currentState.copy(repCount = newRepCount)

        // 检查是否完成当前动作
        if (newRepCount >= currentState.maxReps) {
            completeCurrentExercise()
        }
    }

    /**
     * 完成当前动作
     */
    fun completeCurrentExercise() {
        scoringJob?.cancel()
        _state.value = _state.value.copy(isPlaying = false)

        val score = _state.value.currentScore?.totalScore ?: 0
        val scores = _state.value.exerciseScores + score
        val totalScore = _state.value.totalScore + score

        _state.value = _state.value.copy(
            exerciseScores = scores,
            totalScore = totalScore
        )
    }

    /**
     * 继续下一个动作
     */
    fun nextExercise() {
        val plan = _state.value.plan ?: return
        val nextIndex = _state.value.currentExerciseIndex + 1

        if (nextIndex >= plan.exercises.size) {
            // 计划完成
            _state.value = _state.value.copy(isCompleted = true)
        } else {
            // 下一个动作
            scoringEngine.reset()
            _state.value = _state.value.copy(
                currentExerciseIndex = nextIndex,
                currentScore = null,
                repCount = 0,
                maxReps = plan.exercises[nextIndex].reps,
                isPlaying = false
            )
        }
    }

    /**
     * 更新骨骼数据
     */
    fun updatePoseData(poseData: PoseData) {
        _state.value = _state.value.copy(poseData = poseData)
    }

    /**
     * 暂停运动
     */
    fun pauseExercise() {
        _state.value = _state.value.copy(isPlaying = false)
        scoringJob?.cancel()
    }

    /**
     * 恢复运动
     */
    fun resumeExercise() {
        if (_state.value.currentExercise != null) {
            startExercise()
        }
    }

    /**
     * 重置运动
     */
    fun resetExercise() {
        scoringJob?.cancel()
        scoringEngine.reset()
        previousPoseData = null
        _state.value = ExerciseState()
    }

    override fun onCleared() {
        super.onCleared()
        scoringJob?.cancel()
    }
}

File: ios_native/FamilyMotion/ViewModels/ExerciseViewModel.swift
//
//  ExerciseViewModel.swift
//  FamilyMotion
//
//  Created by FamilyMotion on 2026/06/10.
//

import Foundation
import SwiftUI
import Combine
import AVFoundation
import Vision

// MARK: - 运动视图状态
struct ExerciseViewState {

    // 显示阶段
    enum Phase {
        case preparation      // 准备阶段（倒计时）
        case inProgress       // 运动进行中
        case completed        // 单次动作完成
        case planFinished     // 计划全部完成
    }

    var phase: Phase = .preparation
    var countdown: Int = 3
    var currentScore: ScoreResult? = nil
    var repCount: Int = 0
    var currentExerciseIndex: Int = 0
    var exerciseScores: [Int] = []
    var totalScore: Int {
        exerciseScores.reduce(0, +)
    }
    var isPlaying: Bool = false
}

// MARK: - ExerciseViewModel
@MainActor
final class ExerciseViewModel: ObservableObject {

    // MARK: 状态
    @Published var state: ExerciseViewState = ExerciseViewState()
    @Published var poseData: PoseData? = nil
    @Published var isCameraAuthorized: Bool = false
    @Published var showCameraPermissionAlert: Bool = false

    // MARK: 外部依赖
    let plan: ExercisePlan
    let poseService: PoseDetectionService
    let scoringEngine: ActionScoringEngine

    // MARK: 定时器
    private var timer: Timer?
    private var countdownTimer: Timer?

    // MARK: 初始化
    init(plan: ExercisePlan) {
        self.plan = plan
        self.poseService = PoseDetectionService()
        self.scoringEngine = ActionScoringEngine()

        // 订阅骨骼检测结果
        // 注意：因为我们在不同线程处理，这里手动在外部通过 updatePoseData 传递

        // 检查权限
        Task {
            await checkCameraPermission()
        }
    }

    deinit {
        stopAll()
    }

    // MARK: - 权限检查
    func checkCameraPermission() async {
        let authorized = await poseService.checkCameraPermission()
        isCameraAuthorized = authorized
    }

    // MARK: - 启动倒计时
    func startCountdown() {
        scoringEngine.reset()
        state.countdown = 3
        state.phase = .preparation
        state.repCount = 0

        countdownTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] timer in
            Task { @MainActor in
                guard let self = self else { return }

                if self.state.countdown > 1 {
                    self.state.countdown -= 1
                } else {
                    timer.invalidate()
                    self.startExercise()
                }
            }
        }
    }

    // MARK: - 启动运动
    private func startExercise() {
        state.phase = .inProgress
        state.isPlaying = true

        // 启动相机追踪
        poseService.startTracking()

        // 启动定时评分（每 0.5 秒）
        timer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] timer in
            Task { @MainActor in
                self?.performScoring()
            }
        }
    }

    // MARK: - 执行评分
    private func performScoring() {
        guard let pose = poseData else { return }

        let currentExercise = plan.exercises[state.currentExerciseIndex]
        let result = scoringEngine.scoreAction(
            currentPose: pose,
            targetExercise: currentExercise
        )

        state.currentScore = result

        // 检测是否完成一次动作
        if let previousPose = poseData, pose != previousPose {
            if scoringEngine.detectRepCompletion(
                currentPose: pose,
                previousPose: previousPose,
                exercise: currentExercise
            ) {
                state.repCount += 1

                if state.repCount >= currentExercise.reps {
                    completeCurrentExercise()
                }
            }
        }
    }

    // MARK: - 更新骨骼数据
    func updatePoseData(_ pose: PoseData) {
        self.poseData = pose
    }

    // MARK: - 暂停
    func pauseExercise() {
        state.isPlaying = false
        timer?.invalidate()
        timer = nil
        poseService.stopTracking()
    }

    // MARK: - 恢复
    func resumeExercise() {
        state.isPlaying = true
        poseService.startTracking()

        timer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] timer in
            Task { @MainActor in
                self?.performScoring()
            }
        }
    }

    // MARK: - 完成当前动作
    func completeCurrentExercise() {
        pauseExercise()

        if let score = state.currentScore {
            state.exerciseScores.append(score.totalScore)
        } else {
            state.exerciseScores.append(0)
        }

        state.phase = .completed
    }

    // MARK: - 进入下一个动作
    func nextExercise() {
        if state.currentExerciseIndex + 1 >= plan.exercises.count {
            // 全部完成
            state.phase = .planFinished
        } else {
            state.currentExerciseIndex += 1
            state.repCount = 0
            state.currentScore = nil
            startCountdown()
        }
    }

    // MARK: - 重置
    func reset() {
        stopAll()
        state = ExerciseViewState()
    }

    // MARK: - 停止所有
    private func stopAll() {
        timer?.invalidate()
        countdownTimer?.invalidate()
        timer = nil
        countdownTimer = nil
        poseService.stopTracking()
        state.isPlaying = false
    }

    // MARK: - 计算属性
    var currentExercise: Exercise {
        return plan.exercises[state.currentExerciseIndex]
    }

    var progressText: String {
        return "\(state.currentExerciseIndex + 1) / \(plan.exercises.count)"
    }

    var averageScore: Int {
        guard !state.exerciseScores.isEmpty else { return 0 }
        return state.exerciseScores.reduce(0, +) / state.exerciseScores.count
    }
}
